File: Android/core/src/main/java/com/khushi/conduit/core/protocol/BinaryMessages.kt
package com.khushi.conduit.core.protocol

import java.nio.BufferUnderflowException
import java.nio.ByteBuffer

/**
 * Bodies for the binary channels. Spec: Shared/Protocol/README.md §7.
 * Input is binary rather than JSON because of rate: a trackpad drag is a
 * stream of 10-byte frames, and JSON would multiply that for no benefit.
 */
enum class InputType(val value: Int) {
    POINTER_MOVE(0),
    POINTER_BUTTON(1),
    SCROLL(2),
    KEY(3),
}

enum class PointerButton(val value: Int) { LEFT(0), RIGHT(1), MIDDLE(2) }

/** A physical control on the phone, by intent rather than Android keycode. */
enum class PhoneKey(val value: Int) { VOLUME_UP(0), VOLUME_DOWN(1) }

sealed class InputEvent {
    /** Relative movement: the phone cannot know the Mac's desktop size. */
    data class Move(val dx: Short, val dy: Short) : InputEvent()
    data class Button(val button: PointerButton, val down: Boolean) : InputEvent()
    data class Scroll(val dx: Short, val dy: Short) : InputEvent()
    data class Key(val key: PhoneKey, val down: Boolean) : InputEvent()

    fun toFrame(): Frame = when (this) {
        is Move -> Frame(LinkChannel.INPUT, InputType.POINTER_MOVE.value,
            ByteBuffer.allocate(4).putShort(dx).putShort(dy).array())
        is Button -> Frame(LinkChannel.INPUT, InputType.POINTER_BUTTON.value,
            byteArrayOf(button.value.toByte(), if (down) 1 else 0))
        is Scroll -> Frame(LinkChannel.INPUT, InputType.SCROLL.value,
            ByteBuffer.allocate(4).putShort(dx).putShort(dy).array())
        is Key -> Frame(LinkChannel.INPUT, InputType.KEY.value,
            ByteBuffer.allocate(3).putShort(key.value.toShort()).put(if (down) 1 else 0).array())
    }

    companion object {
        /** Decode an input frame; null for a malformed body rather than a crash. */
        fun from(frame: Frame): InputEvent? {
            if (frame.channel != LinkChannel.INPUT) return null
            val body = ByteBuffer.wrap(frame.payload)
            return try {
                when (frame.type) {
                    InputType.POINTER_MOVE.value -> Move(body.short, body.short)
                    InputType.POINTER_BUTTON.value -> {
                        // Read before matching: a read inside the predicate
                        // would consume a byte per candidate.
                        val raw = body.get().toInt()
                        val button = PointerButton.entries.firstOrNull { it.value == raw } ?: return null
                        Button(button, body.get().toInt() != 0)
                    }
                    InputType.SCROLL.value -> Scroll(body.short, body.short)
                    InputType.KEY.value -> {
                        val raw = body.short.toInt()
                        val key = PhoneKey.entries.firstOrNull { it.value == raw } ?: return null
                        Key(key, body.get().toInt() != 0)
                    }
                    else -> null
                }
            } catch (_: BufferUnderflowException) {
                null
            }
        }
    }
}

object HapticType {
    const val VIBRATE = 0
}

data class Vibration(val durationMs: Int, val amplitude: Int) {
    fun toFrame(): Frame = Frame(LinkChannel.HAPTIC, HapticType.VIBRATE,
        ByteBuffer.allocate(3).putShort(durationMs.toShort()).put(amplitude.coerceIn(1, 255).toByte()).array())

    companion object {
        fun from(frame: Frame): Vibration? {
            if (frame.channel != LinkChannel.HAPTIC || frame.type != HapticType.VIBRATE) return null
            if (frame.payload.size < 3) return null
            val body = ByteBuffer.wrap(frame.payload)
            return Vibration(body.short.toInt() and 0xFFFF, body.get().toInt() and 0xFF)
        }
    }
}

File: Shared/ConduitKit/Sources/ConduitProtocol/BinaryMessages.swift
//
//  BinaryMessages.swift
//  ConduitProtocol
//
//  Bodies for the binary channels. Spec: Shared/Protocol/README.md §7.
//
//  Input is binary rather than JSON because of rate: a trackpad drag at
//  120 Hz is a stream of 10-byte frames, and JSON would multiply that
//  several times over for no benefit.
//

import Foundation

// MARK: - Input (phone → Mac)

public enum InputType: UInt8, Sendable, CaseIterable {
    case pointerMove = 0
    case pointerButton = 1
    case scroll = 2
    case key = 3
}

public enum PointerButton: UInt8, Sendable {
    case left = 0
    case right = 1
    case middle = 2
}

/// A physical control on the phone, described by intent rather than by
/// Android keycode — the Mac decides what it does.
public enum PhoneKey: UInt16, Sendable {
    case volumeUp = 0
    case volumeDown = 1
}

public enum InputEvent: Sendable, Equatable {
    /// Relative movement: the phone cannot know the Mac's desktop size.
    case move(dx: Int16, dy: Int16)
    case button(PointerButton, down: Bool)
    case scroll(dx: Int16, dy: Int16)
    case key(PhoneKey, down: Bool)

    public var frame: Frame {
        var w = ByteWriter()
        let type: InputType
        switch self {
        case let .move(dx, dy):
            type = .pointerMove; w.i16(dx); w.i16(dy)
        case let .button(button, down):
            type = .pointerButton; w.u8(button.rawValue); w.u8(down ? 1 : 0)
        case let .scroll(dx, dy):
            type = .scroll; w.i16(dx); w.i16(dy)
        case let .key(key, down):
            type = .key; w.u16(key.rawValue); w.u8(down ? 1 : 0)
        }
        return Frame(channel: .input, type: type.rawValue, payload: w.data)
    }

    /// Decode an input frame. Returns nil for a malformed body.
    public init?(frame: Frame) {
        guard frame.channel == .input, let type = InputType(rawValue: frame.type) else { return nil }
        var r = ByteReader(frame.payload)
        switch type {
        case .pointerMove:
            guard let dx = r.i16(), let dy = r.i16() else { return nil }
            self = .move(dx: dx, dy: dy)
        case .pointerButton:
            guard let raw = r.u8(), let button = PointerButton(rawValue: raw), let down = r.u8() else { return nil }
            self = .button(button, down: down != 0)
        case .scroll:
            guard let dx = r.i16(), let dy = r.i16() else { return nil }
            self = .scroll(dx: dx, dy: dy)
        case .key:
            guard let raw = r.u16(), let key = PhoneKey(rawValue: raw), let down = r.u8() else { return nil }
            self = .key(key, down: down != 0)
        }
    }
}

// MARK: - Haptic (Mac → phone)

public enum HapticType: UInt8, Sendable {
    case vibrate = 0
}

public struct Vibration: Sendable, Equatable {
    public var durationMs: UInt16
    /// 1–255.
    public var amplitude: UInt8

    public init(durationMs: UInt16, amplitude: UInt8) {
        self.durationMs = durationMs
        self.amplitude = max(amplitude, 1)
    }

    public var frame: Frame {
        var w = ByteWriter()
        w.u16(durationMs)
        w.u8(amplitude)
        return Frame(channel: .haptic, type: HapticType.vibrate.rawValue, payload: w.data)
    }

    public init?(frame: Frame) {
        guard frame.channel == .haptic, frame.type == HapticType.vibrate.rawValue else { return nil }
        var r = ByteReader(frame.payload)
        guard let duration = r.u16(), let amplitude = r.u8() else { return nil }
        self.init(durationMs: duration, amplitude: amplitude)
    }
}

// MARK: - Big-endian helpers

struct ByteWriter {
    private(set) var data = Data()
    mutating func u8(_ v: UInt8) { data.append(v) }
    mutating func u16(_ v: UInt16) {
        data.append(UInt8(truncatingIfNeeded: v >> 8))
        data.append(UInt8(truncatingIfNeeded: v))
    }
    mutating func i16(_ v: Int16) { u16(UInt16(bitPattern: v)) }
}

/// Every read is bounds-checked and returns nil rather than trapping: bytes
/// come off a socket, and a malformed body must fail the message, not the
/// process.
struct ByteReader {
    private let data: Data
    private var offset: Int

    init(_ data: Data) {
        self.data = data
        self.offset = data.startIndex
    }

    mutating func u8() -> UInt8? {
        guard offset < data.endIndex else { return nil }
        defer { offset += 1 }
        return data[offset]
    }

    mutating func u16() -> UInt16? {
        guard offset + 2 <= data.endIndex else { return nil }
        defer { offset += 2 }
        return UInt16(data[offset]) << 8 | UInt16(data[offset + 1])
    }

    mutating func i16() -> Int16? {
        u16().map { Int16(bitPattern: $0) }
    }
}
